crates/sui-storage/src/event_store/sql.rs

```diff
@@ -0,0 +1,513 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//! SQL and SQLite-based Event Store
+
+use super::*;
+
+use async_trait::async_trait;
+use serde_json::{json, Value};
+
+use sqlx::{sqlite::SqliteRow, Executor, Row, SqlitePool};
+use sui_types::event::Event;
+use tracing::{debug, info};
+
+/// Maximum number of events one can ask for right now
+const MAX_LIMIT: usize = 5000;
+
+/// Sqlite-based Event Store
+///
+/// ## Data Model
+/// - Main columns hold most common fields
+/// - object_id is used for multiple purposes, including the Publish package ID
+/// - event_type is an integer in order to save space and corresponds to EventType discriminant
+/// - fields is JSON for now (for easy JSON filtering) and contains all fields not in main columns
+pub struct SqlEventStore {
+    pool: SqlitePool,
+}
+
+const SQL_TABLE_CREATE: &str = "\
+    CREATE TABLE IF NOT EXISTS events(
+        timestamp INTEGER NOT NULL,
+        checkpoint INTEGER,
+        tx_digest BLOB,
+        event_type INTEGER,
+        package_id BLOB,
+        module_name TEXT,
+        object_id BLOB,
+        fields TEXT
```

Comment:
Isn't that just a (package_id, module_name, func_name) triple?    We already have package_id and module_name columns, so I think adding function_name should suffice.  

BTW it is more expensive to index the entire "MoveEventType" triple than individual fields, because the cardinality of a combined field is MUCH higher.   There are comparatively much fewer package IDs and module names.   So indexing three columns gives you much smaller indexes and much more flexibility.  Although for matching that exact string, obviously having just a single column is slightly faster.  Not sure if that's worth it, I think having three columns overall is better.  It's also possible to create an index on all three columns, so either way we don't have to create the MoveEventType column.